xzes40/transformer/src/daemon.cpp

```diff
@@ -0,0 +1,144 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+///////////////////////////////////////////////////////////////////////////////
+
+
+///////////////////////////////////////////////////////////////////////////////
+// Filename: daemon.cpp
+// Description: Creattes daemon binary which waits and spawns jobs for
+//   transformations.
+// Mad props to:
+// - https://github.com/troydhanson/network/blob/master/unixdomain/srv.c
+///////////////////////////////////////////////////////////////////////////////
+
+#include <stdio.h>
+#include <unistd.h>
+#include <sys/socket.h>
+#include <sys/un.h>
+#include <stdlib.h>
+#include <csignal>
+
+#include <lib.hpp>
+#include <transform.hpp>
+#include <daemon.hpp>
+
+#define BUFFER_SIZE 1024
+
+// TODO: CHANGE THIS TO /var/run/... in the final release
+const char *socket_path = "/tmp/xzes40daemon.socket";
+
+int main(int argc, char* argv[]) {
+    int fd;
+
+    signal(SIGINT, xzes::signal_handler); 
+
+    if ((fd = xzes::setup_connection()) == -1)
+        exit(EXIT_FAILURE);
+
+    xzes::daemon(fd);
+
+    return 0;
+}
+
+int xzes::daemon(int fd)
+{
+    char buf[BUFFER_SIZE];
+    int cl, rc;
+
+    for (;;) {
+        if ( (cl = accept(fd, NULL, NULL)) == -1) {
+            perror("XZES40:: Accept error in transformer daemon.");
+            continue;
+        }
+
+        while ( (rc = read(cl, buf, sizeof(buf))) > 0) {
+            if (xzes::valid_request(buf))
+            {
+                printf("XZES40:: Read input `%s`.\n", buf);
+                xzes::cli_arguments_t *args = xzes::parse_request( buf );
+                if (xzes::dispatch_transform( args ))
+                    printf("XZES40:: Processed jobs successfully.\n");
+                else
+                    printf("XZES40:: Failed jobs.\n");
+            }
+        }
+        if (rc == -1) {
+            perror("XZES40:: Read error.\n");
+            exit(EXIT_FAILURE);
+        } else if (rc == 0) {
+            printf("XZES40:: Recieved EOF. Exiting.\n");
+            close(cl);
+        }
+    }
+
+}
+
+int xzes::dispatch_transform( xzes::cli_arguments_t *args )
+{
+    int status = SUCCESS;
+
+    return status;
+}
+
+
+// ------------------------------------------------------------------------
+// Handles signals.
+// Only handles SIGINT by deleting socket file.
+// ------------------------------------------------------------------------
+void xzes::signal_handler(int signum)
+{
+    printf("Recieved signal (%d).", signum);
+
+    if (signum == SIGINT)
+        remove(socket_path);
+
+    exit(signum);
+}
+
+// ----------------------------------------------------------------------------
+// Sets up a unix socket connection and returns the file descriptor.
+// ----------------------------------------------------------------------------
+int xzes::setup_connection()
```

Comment:
If you look at the example daemon and client code you will see where I got most of this code. It's pretty standard unix socket setup code.